lib/components/base_component.dart

```diff
@@ -0,0 +1,147 @@
+  final EffectsHandler _effectsHandler = EffectsHandler();
+
+  final OrderedSet<Component> children =
+      OrderedSet(Comparing.on((c) => c.priority));
+
+  /// This is set by the BaseGame to tell this component to render additional debug information,
+  /// like borders, coordinates, etc.
+  /// This is very helpful while debugging. Set your BaseGame debugMode to true.
+  /// You can also manually override this for certain components in order to identify issues.
+  bool debugMode = false;
+
+  Color get debugColor => const Color(0xFFFF00FF);
+
+  Paint get debugPaint => Paint()
+    ..color = debugColor
+    ..strokeWidth = 1
+    ..style = PaintingStyle.stroke;
+
+  TextConfig get debugTextConfig => TextConfig(color: debugColor, fontSize: 12);
+
+  /// This method is called periodically by the game engine to request that your component updates itself.
+  ///
+  /// The time [dt] in seconds (with microseconds precision provided by Flutter) since the last update cycle.
+  /// This time can vary according to hardware capacity, so make sure to update your state considering this.
+  /// All components on [BaseGame] are always updated by the same amount. The time each one takes to update adds up to the next update cycle.
+  @mustCallSuper
+  @override
+  void update(double dt) {
+    _effectsHandler.update(dt);
+  }
+
+  @mustCallSuper
+  @override
+  void render(Canvas canvas) {
+    prepareCanvas(canvas);
+
+    if (debugMode) {
+      renderDebugMode(canvas);
+    }
+
+    canvas.save();
+    children.forEach((c) => _renderChild(canvas, c));
+    canvas.restore();
+  }
+
+  void renderDebugMode(Canvas canvas) {}
+
+  void _renderChild(Canvas canvas, Component c) {
+    if (!c.loaded) {
+      return;
+    }
+    c.render(canvas);
+    canvas.restore();
+    canvas.save();
+  }
+
+  @protected
+  void prepareCanvas(Canvas canvas) {}
+
+  @mustCallSuper
+  @override
+  void onGameResize(Vector2 gameSize) {
+    super.onGameResize(gameSize);
+    children.forEach((child) => child.onGameResize(gameSize));
+  }
+
+  /// Called to check whether the point is to be counted as within the component
+  /// It needs to be overridden to have any effect, like it is in the
+  /// [PositionComponent]
+  bool checkOverlap(Vector2 point) => false;
+
+  /// Add an effect to the component
+  void addEffect(ComponentEffect effect) {
+    _effectsHandler.add(effect, this);
+  }
+
+  /// Mark an effect for removal on the component
+  void removeEffect(ComponentEffect effect) {
+    _effectsHandler.removeEffect(effect);
+  }
+
+  /// Remove all effects
+  void clearEffects() {
+    _effectsHandler.clearEffects();
+  }
+
+  /// Get a list of non removed effects
+  List<ComponentEffect> get effects => _effectsHandler.effects;
+
+  /// Uses the game passed in to prepare the child component before it is added
+  /// to the list of children
+  void addChild(Game gameRef, Component c) {
+    if (gameRef is BaseGame) {
+      gameRef.prepare(c);
+    }
+    children.add(c);
```

Comment:
I am still concerned about this field being public, I have commented about this on the previous version of the PR and suggested to add a method to iterate the children, without making it public, what do you think?

Reply:
Yeah, I agree.